### PR TITLE
vectorize ortho_group.rvs

### PR DESCRIPTION
**Pull request by Gattocrucco**
#### Reference issue

#16344

#### What does this implement/fix?

New much faster vectorized implementation of `stats.ortho_group.rvs`, the current one loops in python both over the samples and over the rows of each matrix. The algorithm is from the reference already mentioned in the docstring.

#### Additional information

I did not run the tests locally because the scipy build system fails with some error. I have not checked which tests are implemented for this method.

### CODE REVIEW

**Comment by Gattocrucco** [comment]
I managed to run the tests locally and fix them.

The failing ones were reproducibility tests, checking that the random generator with a given seed produces a fixed output. Of course, having changed the algorithm, reproducibility is broken w.r.t. the old version. Is there a policy about this? I think the new implementation is enough of an improvement to justify a backward compatibility break.

**Comment by mdhaber** [comment]
The policy for RNG reproducibility is set by https://numpy.org/neps/nep-0019-rng-policy.html . I'll let you form an opinion before adding mine here.

**Comment by rkern** [comment]
NEP 19 only applies to `numpy.random`. `scipy.stats` has no corresponding policy, though the portion of NEP 19 that you quoted is a reasonable one to adopt for `scipy.stats`, IMO.

**Comment by mdhaber** [comment]
Of course; I meant that it is the the de facto policy in the absence of a SciPy-specific one. I think the approach we've taken in the past is consistent with the NEP.

**Comment by Gattocrucco** [comment]
> One example of a worthwhile improvement would be to change algorithms for a significant increase in performance, for example, moving from the Box-Muller transform method of Gaussian variate generation to the faster Ziggurat algorithm.

Reading this, I guess that this change would be a NEP-19-level improvement, and thus ok.

**Comment by Gattocrucco** [comment]
I see different tests are failing now. The problem is that the tests use numpy 1.21, and `np.linalg.qr` was vectorized in 1.22.

But I wonder why the tests are changing numpy version from one run to the other.

**Comment by rkern** [comment]
The current minimum version of numpy that scipy depends on is 1.19.5. You can only use features available in that release.

**Comment by Gattocrucco** [comment]
Done, although of course I thought it was better to use the vectorized version if present. Reasons:

- I expect many users to have the last version of numpy (common when using venvs)
- Numpy releases every 6 months so I guess it's 1-1.5 years of time before 1.22 is the oldest supported version
- For small matrices it's the vectorization that makes the difference

But I understand the bureaucratic needs of a large project.

**Comment by ilayn** [comment]
By vectorization, do we mean QR accepting nd arrays for n>2?

**Comment by Gattocrucco** [comment]
Yes

**Comment by mdhaber** [comment]
It's not so much the project; it's just what I'm able to review right now. I'll take a closer look today and probably merge as-is.

File: scipy/stats/_multivariate.py [hunk 1]
```diff
@@ -3589,18 +3589,11 @@ def rvs(self, dim, size=1, random_state=None):
...
-            x[0] += D * np.sqrt(norm2)
-            x /= np.sqrt((norm2 - x0**2 + x[0]**2) / 2.)
-            # Householder transformation
-            H[:, n:] = -D * (H[:, n:] - np.outer(np.dot(H[:, n:], x), x))
-        return H
+        z = random_state.normal(size=(dim, dim))
+        q, r = scipy.linalg.qr(z)
+        d = r.diagonal()
+        q *= d/abs(d)
+        return q
```
Comment (mdhaber) [inline scipy/stats/_multivariate.py:3596]: This is a copy-paste of the code from `unitary_group`, the only difference being the initialization of `z`. This is correct according to the discussion in gh-16560.

**Review by mdhaber — APPROVED** [review]
LGTM per the discussion in gh-16560. I also checked that the determinants are -1 or 1 and with equal probability, and the rotation angles (if we interpret the matrix as a rotation) are uniformly distributed.

Thanks @Gattocrucco! You're welcome to submit a follow-up that vectorizes over samples, too; it just might take longer to review, so I just wanted to make sure we got this in.

We need to add this to the release notes - both the enhancements, and the RVS stream-incompatibility. The 1.10.0 release notes disappeared from the Wiki, though, so we need to get that back first....

**Pull request merged**